File: apple/Harness/Services/SwappableFeatureFlagService.swift
//
//  Copyright 2022 Harness
//

import Foundation

class SwappableFeatureFlagService: FeatureFlagService {
    @Inject private var harnessService: HarnessFeatureFlagService
    @Inject private var mockService: MockFeatureFlagService
    
    var useRealService = true
    
    func load(_ callback: @escaping () -> ()) {
        mockService.load { }
        harnessService.load(callback)
    }
    
    func destroy() {
        mockService.destroy()
        harnessService.destroy()
    }
    
    func boolVariation(evaluationId: String, _ callback: @escaping (Bool) -> ()) {
        if (self.useRealService) {
            harnessService.boolVariation(evaluationId: evaluationId, callback)
        } else {
            mockService.boolVariation(evaluationId: evaluationId, callback)
        }
    }
    
    func stringVariation(evaluationId: String, _ callback: @escaping(_ result: String)->()) {
        if (self.useRealService) {
            harnessService.stringVariation(evaluationId: evaluationId, callback)
        } else {
            mockService.stringVariation(evaluationId: evaluationId, callback)
        }
    }
    func numberVariation(evaluationId: String, _ callback: @escaping (Int) -> ()) {
        if (self.useRealService) {
            harnessService.numberVariation(evaluationId: evaluationId, callback)
        } else {
            mockService.numberVariation(evaluationId: evaluationId, callback)
        }
    }
    
    func jsonVariation(evaluationId: String, _ callback: @escaping (String) -> ()) {
        if (self.useRealService) {
            harnessService.jsonVariation(evaluationId: evaluationId, callback)
        } else {
            mockService.jsonVariation(evaluationId: evaluationId, callback)
        }
    }
}

File: android/app/src/main/java/io/harness/services/SwappableFeatureFlagService.kt
package io.harness.services

import io.harness.HarnessService
import io.harness.MockService
import io.harness.settings.SettingsRepository
import io.harness.settings.SettingsRepository.Companion.SETTING_USE_REAL_SERVICE
import javax.inject.Inject

class SwappableFeatureFlagService @Inject constructor(
    @HarnessService private val harnessFeatureFlagService: FeatureFlagService,
    @MockService private val mockFeatureFlagService: FeatureFlagService,
    private val settingsRepository: SettingsRepository
) : FeatureFlagService {
    override fun load(callback: () -> Unit) {
        harnessFeatureFlagService.load(callback)
        mockFeatureFlagService.load {}
    }

    override fun destroy() {
        harnessFeatureFlagService.destroy()
        mockFeatureFlagService.destroy()
    }

    override fun boolVariation(evaluationId: String, callback: FeatureFlagCallback<Boolean>) =
        if (useRealService) {
            harnessFeatureFlagService.boolVariation(evaluationId, callback)
        } else {
            mockFeatureFlagService.boolVariation(evaluationId, callback)
        }

    override fun stringVariation(evaluationId: String, callback: FeatureFlagCallback<String>) =
        if (useRealService) {
            harnessFeatureFlagService.stringVariation(evaluationId, callback)
        } else {
            mockFeatureFlagService.stringVariation(evaluationId, callback)
        }

    override fun numberVariation(evaluationId: String, callback: FeatureFlagCallback<Number>) =
        if (useRealService) {
            harnessFeatureFlagService.numberVariation(evaluationId, callback)
        } else {
            mockFeatureFlagService.numberVariation(evaluationId, callback)
        }

    override fun jsonVariation(evaluationId: String, callback: FeatureFlagCallback<JSONString>) =
        if (useRealService) {
            harnessFeatureFlagService.jsonVariation(evaluationId, callback)
        } else {
            mockFeatureFlagService.jsonVariation(evaluationId, callback)
        }

    private val useRealService: Boolean
        get() = settingsRepository.get(SETTING_USE_REAL_SERVICE, "true").toBoolean()
}